File: mvp-cli/modules/metrics.py
"""Module for displaying performance metrics of the code."""
from typing import Dict, Optional
import time
from modules.formatting import BOLD_STYLE, NORMAL_STYLE, DIM_STYLE, CARET
import modules.config as config

# Maps names to their start times
_metrics_start_map: Dict[str, float] = dict()


def start(name: str, message: str):
  """
  Starts a metric, printing out the message indicating that it was started.

  Params:
    `name`: the name of the metric to start
    `message`: the message to print out after starting the metric
  """
  if not config.debug_show_metrics:
    return

  print(f"\n{CARET} {NORMAL_STYLE}{message}...", end='', flush=True)

  _metrics_start_map[name] = time.time()


def end(name: str, suffix: Optional[str] = None):
  """
  Ends a metric, printing out "done!" and the time the metric took to complete.

  Params:
    `name`: the name of the metric to end
    `suffix`: an optional suffix string to print out
  """
  if not config.debug_show_metrics:
    return

  end_time = time.time()
  start_time = _metrics_start_map.get(name, None)

  if start_time is None:
    raise f"Error: metric '{name}' was never started"

  del _metrics_start_map[name]

  total_time = round(end_time - start_time, 4)
  suffix_str = f"{NORMAL_STYLE}({suffix})" if suffix is not None else ""

  print(f" {BOLD_STYLE}done! {suffix_str} {DIM_STYLE}({total_time}s)\n", flush=True)

File: core-lib/src/metrics.cpp
#include "metrics.hpp"
#include "formatting.hpp"
#include <format>
#include <iostream>

namespace shabam {
Metrics::Metrics() {};

void Metrics::start(std::string name, std::string message) {
  this->start(name, message, false);
};

void Metrics::start(std::string name, std::string message,
                    bool include_newline) {
  std::cout << formatting::caret << " " << formatting::normal_style << message
            << "... " << formatting::reset_style;

  if (include_newline) {
    std::cout << std::endl;
  }

  const auto start_time = std::chrono::high_resolution_clock::now();
  this->start_map[name] = start_time;
}

void Metrics::end(std::string name) { this->end(name, ""); };

void Metrics::end(std::string name, std::string suffix) {
  const auto end_time = std::chrono::high_resolution_clock::now();

  if (!this->start_map.contains(name)) {
    throw std::invalid_argument(
        std::format("Metric \"{}\" was never started", name));
  }

  const auto start_time = this->start_map[name];

  // Remove item from start map
  auto iter = this->start_map.find(name);
  if (iter != this->start_map.end()) {
    this->start_map.erase(iter);
  }

  std::chrono::duration<double, std::milli> duration_ms = end_time - start_time;

  std::string suffix_fmt = suffix != "" ? suffix + " " : "";

  std::cout << formatting::bold_style << "done! " << formatting::normal_style
            << suffix_fmt << formatting::dim_style << "(" << duration_ms << ")"
            << formatting::reset_style << std::endl;
};
} // namespace shabam